libavcodec/libdavs2.c

```diff
@@ -0,0 +1,204 @@
+        frame->buf[i]      = av_buffer_alloc(size_plane);
+        frame->data[i]     = frame->buf[i]->data;
+        frame->linesize[i] = pic->width[i] * bytes_per_sample;
+        if (!frame->buf[i] || !frame->data[i] || !frame->linesize[i]){
+            av_log(avctx, AV_LOG_ERROR, "dump error: alloc failed.\n");
+            return AVERROR(EINVAL);
+        }
+        memcpy(frame->data[i], pic->planes[i], size_plane);
+    }
+
+    frame->width     = cad->headerset.horizontal_size;
+    frame->height    = cad->headerset.vertical_size;
+    frame->pts       = cad->out_frame.pts;
+    frame->pict_type = pic->type;
+    frame->format    = avctx->pix_fmt;
+
+    cad->decoded_frames++;
+    return 1;
+}
+
+static av_cold davs2_end(AVCodecContext *avctx)
+{
+    DAVS2Context *cad = avctx->priv_data;
+
+    /* close the decoder */
+    if (cad->decoder) {
+        davs2_decoder_close(cad->decoder);
+        av_log(avctx, AV_LOG_VERBOSE, "decoder destroyed. %p; frames %d\n", cad->decoder, cad->decoded_frames);
+        cad->decoder = NULL;
+    }
+
+    return 0;
+}
+
+static int davs2_decode_frame(AVCodecContext *avctx, void *data, int *got_frame, AVPacket *avpkt)
+{
+    DAVS2Context *cad = avctx->priv_data;
+    int buf_size       = avpkt->size;
+    uint8_t *buf_ptr = avpkt->data;
+    AVFrame *frame = data;
+    int ret = 0;
+
+    *got_frame = 0;
+    cad->frame = frame;
+    avctx->flags |= AV_CODEC_FLAG_TRUNCATED;
+
+    if (!buf_size) {
+        cad->packet.data = buf_ptr;
+        cad->packet.len  = buf_size;
+        cad->packet.pts  = avpkt->pts;
+        cad->packet.dts  = avpkt->dts;
+
+        while (1) {
+            ret = davs2_decoder_flush(cad->decoder, &cad->headerset, &cad->out_frame);
+
+            if (ret < 0)
+                return 0;
+
+            if (cad->out_frame.ret_type != DAVS2_DEFAULT) {
+                *got_frame = davs_dump_frames(avctx, &cad->out_frame, &cad->headerset, frame);
+                davs2_decoder_frame_unref(cad->decoder, &cad->out_frame);
+            }
+            if (*got_frame)
+                break;
+        }
+        return 0;
+    } else {
+        while (buf_size > 0) {
+            int len = buf_size;   // for API-3, pass all data in
+
+            cad->packet.marker = 0;
+            cad->packet.data = buf_ptr;
+            cad->packet.len  = len;
+            cad->packet.pts  = avpkt->pts;
+            cad->packet.dts  = avpkt->dts;
+
+            len = davs2_decoder_decode(cad->decoder, &cad->packet, &cad->headerset, &cad->out_frame);
+
+            if (cad->out_frame.ret_type != DAVS2_DEFAULT) {
+                *got_frame = davs_dump_frames(avctx, &cad->out_frame, &cad->headerset, frame);
+                davs2_decoder_frame_unref(cad->decoder, &cad->out_frame);
+            }
+
+            if (len < 0) {
+                av_log(avctx, AV_LOG_ERROR, "A decoder error counted\n");
+                if (cad->decoder) {
```

Comment:
duplicate with the mark comment above, maybe define a static function is better?